sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -150,6 +153,17 @@ class DataFrame private[sql](
     }
   }
 
+  /**
+   * Resolves a column name. This is called when it is required to resolve a column by its
+   * name only and not as a column path..
+   */
+  private[sql] def resolveColName(colName: String, userSuppliedName: String): Boolean = {
```

Comment:
how about
```
private[sql] def indexOf(colName: String): Option[Int] = {
  val resolver = sqlContext.analyzer.resolver
  val index = queryExecution.analyzed.output.indexWhere(f => resolver(f.name, colName))
  if (index >= 0) Some(index) else None
}
```

then we can rewrite `withColumn` to:
```
indexOf(colName).map { index =>
  select(output.updated(index, col.as(colName)).map(Column(_)) : _*)
}.getOrElse {
  select(Column("*"), col.as(colName))
}
```

There may be better name for this, like `resolveToIndex`